java-squid/src/main/java/org/sonar/java/se/SymbolicValueRelation.java

```diff
@@ -34,73 +38,154 @@ protected SymbolicValueRelation(SymbolicValue v1, SymbolicValue v2) {
+  }
+
+  private Collection<SymbolicValueRelation> transitiveReduction(Collection<SymbolicValueRelation> knownRelations, Set<SymbolicValueRelation> usedRelations) {
+    boolean changed = false;
+    List<SymbolicValueRelation> result = new ArrayList<>();
+    for (SymbolicValueRelation relation : knownRelations) {
+      boolean used = false;
+      if (v1.equals(relation.v1)) {
+        used = result.addAll(relation.combinedRelations(knownRelations, usedRelations));
+      } else if (v1.equals(relation.v2)) {
+        used = result.addAll(relation.symmetric().combinedRelations(knownRelations, usedRelations));
+      }
+      if (used) {
+        changed = true;
+      } else {
+        result.add(relation);
       }
     }
-    return null;
+    return changed ? result : Collections.<SymbolicValueRelation>emptyList();
   }
 
-  protected boolean hasSameOperand(SymbolicValueRelation relation) {
-    return (v1.equals(relation.v1) && v2.equals(relation.v2))
-      || (v1.equals(relation.v2) && v2.equals(relation.v1));
+  private Collection<SymbolicValueRelation> combinedRelations(Collection<SymbolicValueRelation> relations, Set<SymbolicValueRelation> usedRelations) {
+    List<SymbolicValueRelation> result = new ArrayList<>();
+    for (SymbolicValueRelation relation : relations) {
+      SymbolicValueRelation combined = null;
+      if (v2.equals(relation.v1) && !v1.equals(relation.v2)) {
+        combined = relation.combinedAfter(this);
+      } else if (v2.equals(relation.v2) && !v1.equals(relation.v2)) {
+        combined = relation.symmetric().combinedAfter(this);
+      }
+      if (combined != null && !usedRelations.contains(combined)) {
+        result.add(combined);
+      }
+    }
+    return result;
   }
 
   @Override
   public String toString() {
     StringBuilder buffer = new StringBuilder();
-    buffer.append('(');
     buffer.append(v1);
     buffer.append(' ');
     buffer.append(getOperand());
     buffer.append(' ');
     buffer.append(v2);
-    buffer.append(')');
     return buffer.toString();
   }
 
   protected abstract String getOperand();
 
+  /**
+   * @return a new relation, the negation of the receiver
+   */
   protected abstract SymbolicValueRelation inverse();
 
   /**
+   * @return a new relation, equivalent to the receiver with inverted parameters
+   */
+  protected abstract SymbolicValueRelation symmetric();
+
+  /**
    * @param a relation between symbolic values
    * @return <ul>
-   * <li>TRUE  if the supplied relation implies that the receiver is true</li>
-   * <li>FALSE if the supplied relation implies that the receiver is false</li>
+   * <li>TRUE  if the receiver implies that the supplied relation is true</li>
+   * <li>FALSE if the receiver implies that the supplied relation is false</li>
    * <li>null  otherwise</li>
    * </ul>
    */
-  protected abstract Boolean implies(SymbolicValueRelation relation);
+  @CheckForNull
+  public Boolean implies(SymbolicValueRelation relation) {
```

Comment:
To be able to test it... As the number of relations increase, it will be essential to write all possible tests.